### PR TITLE
Setup for using sql models

### PR DESCRIPTION
**Pull request by benrudolph**
@snopoke this i think sets the stage for using sql models and incorporates some of the refactors you mentioned. i think we're  just down to `post_xform` and `submit_form_locally` as the two ways to save xforms which is pretty cool

cc: @esoergel

### CODE REVIEW

File: corehq/ex-submodules/couchforms/tests/test_archive.py [hunk 1]
```diff
@@ -1,27 +1,43 @@
+import io
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_archive.py]: F401 'io' imported but unused

File: corehq/form_processor/utils.py [hunk 2]
```diff
@@ -1,6 +1,10 @@
+from django.conf import settings
 import functools
 import types
 import collections
+from corehq.toggles import USE_SQL_BACKEND
+
+from django.conf import settings
```
Comment (dimagimon) [inline corehq/form_processor/utils.py]: F811 redefinition of unused 'settings' from line 1

File: corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py [hunk 3]
```diff
@@ -133,12 +133,11 @@ def testSubmitToDeletedCase(self):
         deleted_doc_type = 'CommCareCase-Deleted'
         [xform, [case]] = FormProcessorInterface.post_case_blocks([
             CaseBlock(create=True, case_id=case_id, user_id='whatever',
-                      update={'foo': 'bar'}).as_xml()
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py]: Why is this necessary - I don't think we should ever change the doc_type via a form submission.

File: corehq/form_processor/backends/couch.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    @staticmethod
+    def get_by_doc_type(domain, doc_type):
+        return XFormError.view(
+            'domain/docs',
+            startkey=[domain, doc_type],
+            endkey=[domain, doc_type, {}],
+            reduce=False,
+            include_docs=True,
+        ).all()
```
Comment (snopoke) [inline corehq/form_processor/backends/couch.py]: Shouldn't this wrap it based on the doc type? Similar to `doc_types()[doc['doc_type']].wrap(doc)`

**Comment by snopoke** [comment]
Nice, this is looking really good.
Test failure:
```
======================================================================
FAIL: test_basic_edit (couchforms.tests.test_edits.EditFormTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/couchforms/tests/test_edits.py", line 51, in test_basic_edit
    self.assertEqual(self.ID, xform.id)
AssertionError: '7H46J37FGH3' != u'495e16d0dbf69a5d96351a9f2e3779db'
```

**Comment by benrudolph** [comment]
cool, was looking into that the other day. the database isn't getting cleaned up properly. will have fix soon

File: corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py (hunk 3 above)
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py]: hmm you're right i was just looking for a quick fix to not have to use `update_properties` to change the doc_type, but seems i could just delete the form and that'd have the same effect

File: corehq/form_processor/backends/couch.py (hunk 4 above)
Comment (benrudolph) [inline corehq/form_processor/backends/couch.py]: :+1:

File: corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py (hunk 3 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py]: You mean delete the case. But yes, that should work.

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py]: oops yep

File: corehq/form_processor/backends/couch.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+from couchdbkit import ResourceNotFound
```
Comment (benrudolph) [inline corehq/form_processor/backends/couch.py]: we also may want to rename this file to `backends/couch/xform.py` if we feel like there is going to be a 1-to-1 mapping from interface functions to backend functions (that's how it currently is right now). if it's not going to be like that then we can just leave as is. this'll probably be something we figure out later down the road -- just flagging for now

File: corehq/form_processor/utils.py [hunk 6]
```diff
@@ -33,3 +35,12 @@ def _inner(*args, **kwargs):
         return _wrap(obj)
 
     return _inner
+
+
+def should_use_sql_backend(domain):
+    toggle_to_check = USE_SQL_BACKEND
```
Comment (esoergel) [inline corehq/form_processor/utils.py]: curious what this is for

File: settings.py [hunk 7]
```diff
@@ -968,6 +968,8 @@
 DATADOG_API_KEY = None
 DATADOG_APP_KEY = None
 
+TEST_SHOULD_USE_SQL_BACKEND = False
+
```
Comment (esoergel) [inline settings.py]: Can you elaborate a bit on this setting?  When do you expect it to be enabled or disabled?

Comment (benrudolph) [inline settings.py]: sure, so i'm gonna answer both your questions here. we want to be able to use the postgres backend in tests. but also keep the ability to run couch tests. this setting will allow us to use `override_settings` to choose some tests to use one or the other. we can then create a decorator to run tests on both settings. see simon's comment here: https://github.com/dimagi/commcare-hq/pull/8780#discussion_r42839279 for more context

File: corehq/form_processor/utils.py (hunk 6 above)
Comment (benrudolph) [inline corehq/form_processor/utils.py]: this is just the toggle so that we can turn the postgres db on at a per domain basis. also this at the moment doesn't need to be a variable, but once it becomes a random toggle we'll need to make sure the randomness is 0 in the tests

File: corehq/ex-submodules/couchforms/tests/test_errors.py [hunk 8]
```diff
@@ -1,7 +1,8 @@
 from django.test import TestCase
 from corehq.apps.receiverwrapper import submit_form_locally
 
-from corehq.form_processor.interfaces.xform import XFormInterface
+from corehq.form_processor.test_utils import FormProcessorTestUtils
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_errors.py]: F401 'FormProcessorTestUtils' imported but unused

File: corehq/form_processor/backends/couch.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+from couchforms.models import doc_types, XFormInstance, XFormError
```
Comment (dimagimon) [inline corehq/form_processor/backends/couch.py]: F401 'XFormError' imported but unused

File: corehq/form_processor/utils.py (hunk 6 above)
Comment (esoergel) [inline corehq/form_processor/utils.py]: I meant the variable part of it specifically, given that it's only used once, and there's no forking on what gets assigned to `toggle_to_check`.

Comment (benrudolph) [inline corehq/form_processor/utils.py]: yep you're right, will update

File: corehq/form_processor/generic.py [hunk 10]
```diff
@@ -201,6 +204,10 @@ def reverse_indices(self):
         from corehq.form_processor.interfaces.case import CaseInterface
         return CaseInterface.get_reverse_indices(self.domain, self.id)
 
+    @property
+    def is_deleted(self):
+        return self.doc_type.endswith(DELETED_SUFFIX)
```
Comment (snopoke) [inline corehq/form_processor/generic.py]: OK for now but should probably be a property that gets set when converting from XFormInstance e.g. https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/couchforms/models.py#L451-451

File: settings.py (hunk 7 above)
Comment (snopoke) [inline settings.py]: I don't think this should be in the main settings file. If can go in travis settings at some point. Also here it is `TEST_` but  where it's referenced it is `TESTS_`

**Comment by snopoke** [comment]
Nothing blocking so merging. Nice work.

**Comment by czue** [comment]
onwards!

File: corehq/form_processor/generic.py (hunk 10 above)
Comment (benrudolph) [inline corehq/form_processor/generic.py]: agreed, fiddled with doing that for a moment and then just went the easy route. marked it down to change though

File: settings.py (hunk 7 above)
Comment (benrudolph) [inline settings.py]: :+1:
